AC should work with pre-dispatch IR

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #164505

Previously we had to rely on turning off export verifier because the AC body was torch IR instead of aten IR. This PR makes it so that we create an IR that is export compatible.

tugsbayasgalan committed:
AC should work with pre-dispatch IR

[ghstack-poisoned]

xmfan commented on torch/_higher_order_ops/wrap.py:
```diff
@@ -319,11 +320,12 @@ def proxy_mode_key(
-            node.meta["val"] = node.meta["example_value"]
+
+    # TODO (tmanlaibaatar) don't we need flat_apply here??
+    flat_args, _ = pytree.tree_flatten((args, kwargs))
```
Yeah, in strict mode export, gmod here is a torch IR graph module. Tracing it into pre-dispatch aten here makes sense to me.

xmfan commented on test/export/test_export.py:
```diff
@@ -717,6 +719,98 @@ def example_inputs(self):
+        with stack:
+            jwd = aot_export_joint_with_descriptors(stack, ep.module(), (x,))
+            self.assertExpectedInline(
+                str(jwd.graph_module.code).strip(),
```
you probably want to have CI guard that the tags are applied properly, something like this for SAC: https://github.com/pytorch/pytorch/blob/f465ea6752c91498de63eb57439a74f4836e568a/test/higher_order_ops/test_local_map.py#L313-L324. Forgot what it is for full AC, maybe "recompute" is MUST_SAVE

xmfan reviewed (approved):
cc @soulitzer

tugsbayasgalan committed:
Update on "AC should work with pre-dispatch IR"


Previously we had to rely on turning off export verifier because the AC body was torch IR instead of aten IR. This PR makes it so that we create an IR that is export compatible. 



[ghstack-poisoned]

tugsbayasgalan commented on torch/_higher_order_ops/wrap.py:
```diff
@@ -312,18 +313,24 @@ def proxy_mode_key(
+    flat_args, _ = pytree.tree_flatten((args, kwargs))
+    with fx_traceback.preserve_node_meta():
+        gmod_aten = reenter_make_fx(Interpreter(gmod).run)(*flat_args)
+        gmod_aten.meta["_checkpoint_context_fn"] = gmod.meta["_checkpoint_context_fn"]
```
cc: @xmfan

xmfan replied:
.meta was a mistake 😭

tugsbayasgalan committed:
Update on "AC should work with pre-dispatch IR"


Previously we had to rely on turning off export verifier because the AC body was torch IR instead of aten IR. This PR makes it so that we create an IR that is export compatible. 



[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D83851604).

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D83851604).

tugsbayasgalan commented:
@pytorchbot label "topic: not user facing"

tugsbayasgalan commented:
@pytorchbot merge -i
